src/Microsoft.AspNetCore.StaticFiles/StaticFileContext.cs

```diff
@@ -346,23 +346,39 @@ public Task SendStatusAsync(int statusCode)
         {
             ApplyResponseHeaders(Constants.Status200Ok);
 
-            string physicalPath = _fileInfo.PhysicalPath;
-            var sendFile = _context.Features.Get<IHttpSendFileFeature>();
-            if (sendFile != null && !string.IsNullOrEmpty(physicalPath))
-            {
-                await sendFile.SendFileAsync(physicalPath, 0, _length, _context.RequestAborted);
-                return;
-            }
-
-            Stream readStream = _fileInfo.CreateReadStream();
             try
             {
-                // Larger StreamCopyBufferSize is required because in case of FileStream readStream isn't going to be buffering
-                await StreamCopyOperation.CopyToAsync(readStream, _response.Body, _length, StreamCopyBufferSize, _context.RequestAborted);
+                string physicalPath = _fileInfo.PhysicalPath;
+                var sendFile = _context.Features.Get<IHttpSendFileFeature>();
+                if (sendFile != null && !string.IsNullOrEmpty(physicalPath))
+                {
+                    await sendFile.SendFileAsync(physicalPath, 0, _length, _context.RequestAborted);
+                    return;
+                }
+
+                Stream readStream = _fileInfo.CreateReadStream();
+                try
+                {
+                    // Larger StreamCopyBufferSize is required because in case of FileStream readStream isn't going to be buffering
+                    await StreamCopyOperation.CopyToAsync(readStream, _response.Body, _length, StreamCopyBufferSize, _context.RequestAborted);
+                }
+                finally
+                {
+                    readStream.Dispose();
+                }
             }
-            finally
+            catch (OperationCanceledException ex)
             {
-                readStream.Dispose();
+                _logger.LogWriteCancelled(ex);
+                if (_context.Response.HasStarted)
```

Comment:
Couldn't `HasStarted` theoretically be replaced just like `RequestAborted`? I think trusting the `RequestAborted` token is reasonable enough. If you don't trust it, I would call `_context.Abort()` in any case.